File: services/api/dto/projects.go
package dto

import "time"

// ── Project list / detail ─────────────────────────────────────────────────────

type ProjectListItem struct {
	ID                   string     `json:"id"`
	Name                 string     `json:"name"`
	Address              *string    `json:"address"`
	Description          *string    `json:"description"`
	Status               string     `json:"status"`
	StartDate            *string    `json:"start_date"`
	EndDate              *string    `json:"end_date"`
	ClosedAt             *time.Time `json:"closed_at"`
	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`
	PrimaryPoslovodaID   *string    `json:"primary_poslovoda_id"`
	PrimaryPoslovodaName *string    `json:"primary_poslovoda_name"`
	AssignmentCount      int        `json:"assignment_count"`
}

type AssignmentItem struct {
	ID            string    `json:"id"`
	EmployeeID    string    `json:"employee_id"`
	EmployeeName  string    `json:"employee_name"`
	EmployeeRole  string    `json:"employee_role"`
	RoleOnProject string    `json:"role_on_project"`
	Active        bool      `json:"active"`
	AssignedAt    time.Time `json:"assigned_at"`
}

type ProjectDetail struct {
	ID                   string           `json:"id"`
	CompanyID            string           `json:"company_id"`
	Name                 string           `json:"name"`
	Address              *string          `json:"address"`
	Description          *string          `json:"description"`
	Status               string           `json:"status"`
	StartDate            *string          `json:"start_date"`
	EndDate              *string          `json:"end_date"`
	ClosedAt             *time.Time       `json:"closed_at"`
	CreatedBy            *string          `json:"created_by"`
	ClosedBy             *string          `json:"closed_by"`
	CreatedAt            time.Time        `json:"created_at"`
	UpdatedAt            time.Time        `json:"updated_at"`
	PrimaryPoslovodaID   *string          `json:"primary_poslovoda_id"`
	PrimaryPoslovodaName *string          `json:"primary_poslovoda_name"`
	MaterialsCount       int              `json:"materials_count"`
	DailyReportsCount    int              `json:"daily_reports_count"`
	Assignments          []AssignmentItem `json:"assignments"`
}

// ── Requests ──────────────────────────────────────────────────────────────────

type CreateProjectRequest struct {
	Name               string  `json:"name" binding:"required"`
	Address            *string `json:"address"`
	Description        *string `json:"description"`
	StartDate          *string `json:"start_date"`
	EndDate            *string `json:"end_date"`
	PrimaryPoslovodaID *string `json:"primary_poslovoda_id"`
}

type UpdateProjectRequest struct {
	Name        string  `json:"name" binding:"required"`
	Address     *string `json:"address"`
	Description *string `json:"description"`
	StartDate   *string `json:"start_date"`
	EndDate     *string `json:"end_date"`
}

type AssignPoslovodaRequest struct {
	PoslovodaID string `json:"poslovoda_id" binding:"required"`
}

// ── Archive list ──────────────────────────────────────────────────────────────

type ArchiveListItem struct {
	ID                     string     `json:"id"`
	Name                   string     `json:"name"`
	Address                *string    `json:"address"`
	Status                 string     `json:"status"`
	StartDate              *string    `json:"start_date"`
	EndDate                *string    `json:"end_date"`
	ClosedAt               *time.Time `json:"closed_at"`
	ClosedByName           *string    `json:"closed_by_name"`
	PrimaryPoslovodaName   *string    `json:"primary_poslovoda_name"`
	DailyReportsCount      int        `json:"daily_reports_count"`
	MaterialPurchasesCount int        `json:"material_purchases_count"`
	ProjectMaterialsCount  int        `json:"project_materials_count"`
}

type ArchiveListResponse struct {
	Projects   []ArchiveListItem `json:"projects"`
	Total      int               `json:"total"`
	Page       int               `json:"page"`
	PerPage    int               `json:"per_page"`
	TotalPages int               `json:"total_pages"`
}

// ── Archive summary ───────────────────────────────────────────────────────────

type ArchiveSummaryAssignee struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Active bool   `json:"active"`
}

type ArchiveSummary struct {
	ID                          string                   `json:"id"`
	Name                        string                   `json:"name"`
	Address                     *string                  `json:"address"`
	Status                      string                   `json:"status"`
	StartDate                   *string                  `json:"start_date"`
	EndDate                     *string                  `json:"end_date"`
	ClosedAt                    *time.Time               `json:"closed_at"`
	ClosedByName                *string                  `json:"closed_by_name"`
	DailyReportsCount           int                      `json:"daily_reports_count"`
	WorkerHoursTotal            float64                  `json:"worker_hours_total"`
	ActivitiesCount             int                      `json:"activities_count"`
	VtkCount                    int                      `json:"vtk_count"`
	PurchaseSessionsCount       int                      `json:"purchase_sessions_count"`
	PurchaseItemsCount          int                      `json:"purchase_items_count"`
	ProjectMaterialsCount       int                      `json:"project_materials_count"`
	MaterialResponsibilityCount int                      `json:"material_responsibility_count"`
	TransfersCount              int                      `json:"transfers_count"`
	AssignedPoslovode           []ArchiveSummaryAssignee `json:"assigned_poslovode"`
}

File: apps/web/lib/types/projects.ts
export type ProjectStatus = 'active' | 'closed' | 'archived'

export interface ProjectListItem {
  id: string
  name: string
  address: string | null
  description: string | null
  status: ProjectStatus
  start_date: string | null
  end_date: string | null
  closed_at: string | null
  created_at: string
  updated_at: string
  primary_poslovoda_id: string | null
  primary_poslovoda_name: string | null
  assignment_count: number
}

export interface ProjectAssignment {
  id: string
  employee_id: string
  employee_name: string
  employee_role: string
  role_on_project: string
  active: boolean
  assigned_at: string
}

export interface ProjectDetail extends ProjectListItem {
  company_id: string
  created_by: string | null
  closed_by: string | null
  materials_count: number
  daily_reports_count: number
  assignments: ProjectAssignment[]
}

export interface CreateProjectPayload {
  name: string
  address?: string | null
  description?: string | null
  start_date?: string | null
  end_date?: string | null
  primary_poslovoda_id?: string | null
}

export interface UpdateProjectPayload {
  name: string
  address?: string | null
  description?: string | null
  start_date?: string | null
  end_date?: string | null
}

export const STATUS_LABELS: Record<ProjectStatus, string> = {
  active: 'Aktivno',
  closed: 'Zatvoreno',
  archived: 'Arhivirano',
}

export function canManageProjects(role: string): boolean {
  return role === 'direktor' || role === 'inzenjer'
}

export function canViewProjects(role: string): boolean {
  return ['direktor', 'inzenjer', 'poslovoda', 'radnik'].includes(role)
}

export function canViewArchive(role: string): boolean {
  return ['direktor', 'inzenjer'].includes(role)
}

// ── Archive list ──────────────────────────────────────────────────────────────

export interface ArchiveListItem {
  id: string
  name: string
  address: string | null
  status: ProjectStatus
  start_date: string | null
  end_date: string | null
  closed_at: string | null
  closed_by_name: string | null
  primary_poslovoda_name: string | null
  daily_reports_count: number
  material_purchases_count: number
  project_materials_count: number
}

export interface ArchiveListResponse {
  projects: ArchiveListItem[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

// ── Archive summary ───────────────────────────────────────────────────────────

export interface ArchiveSummaryAssignee {
  id: string
  name: string
  active: boolean
}

export interface ArchiveSummary {
  id: string
  name: string
  address: string | null
  status: ProjectStatus
  start_date: string | null
  end_date: string | null
  closed_at: string | null
  closed_by_name: string | null
  daily_reports_count: number
  worker_hours_total: number
  activities_count: number
  vtk_count: number
  purchase_sessions_count: number
  purchase_items_count: number
  project_materials_count: number
  material_responsibility_count: number
  transfers_count: number
  assigned_poslovode: ArchiveSummaryAssignee[]
}
